plugins/discourse-presence/assets/javascripts/discourse/lib/presence-manager.js.es6

```diff
@@ -0,0 +1,205 @@
+
+    this.setProperties({
+      users: [],
+      editingUsers: [],
+      subscribed: false
+    });
+  },
+
+  subscribe() {
+    if (this.subscribed) return;
+
+    this.messageBus.subscribe(
+      this.channel,
+      message => {
+        const { user, state } = message;
+        if (this.get("currentUser.id") === user.id) return;
+
+        switch (state) {
+          case REPLYING:
+            this._appendUser(this.users, user);
+            break;
+          case EDITING:
+            this._appendUser(this.editingUsers, user, {
+              post_id: parseInt(message.post_id, 10)
+            });
+            break;
+          case CLOSED:
+            this._removeUser(user);
+            break;
+        }
+      },
+      MESSAGE_BUS_LAST_ID
+    );
+
+    this.set("subscribed", true);
+  },
+
+  unsubscribe() {
+    this.messageBus.unsubscribe(this.channel);
+    this._stopTimer();
+    this.set("subscribed", false);
+  },
+
+  @discourseComputed("topic.id")
+  channel(topicId) {
+    return `/presence/${topicId}`;
+  },
+
+  throttlePublish(state, whisper, postId) {
+    return throttle(
+      this,
+      this.publish,
+      state,
+      whisper,
+      postId,
+      KEEP_ALIVE_DURATION_SECONDS * 1000
+    );
+  },
+
+  publish(state, whisper, postId) {
+    const data = {
+      state,
+      topic_id: this.get("topic.id")
+    };
+
+    if (whisper) {
+      data.is_whisper = 1;
+    }
+
+    if (postId) {
+      data.post_id = postId;
+    }
+
+    return ajax("/presence/publish", {
+      type: "POST",
+      data
+    });
+  },
+
+  _debounceSetProperties(properties) {
+    debounce(this, this.setProperties, properties, 200);
+  },
+
+  _removeUser(user) {
+    [this.users, this.editingUsers].forEach(users => {
+      const existingUser = users.findBy("id", user.id);
+      if (existingUser) users.removeObject(existingUser);
+    });
+  },
+
+  _cleanUpUsers() {
+    [this.users, this.editingUsers].forEach(users => {
+      const staleUsers = [];
+
+      users.forEach(user => {
+        if (user.last_seen <= Date.now() - BUFFER_DURATION_SECONDS * 1000) {
+          staleUsers.push(user);
+        }
+      });
+
+      users.removeObjects(staleUsers);
+    });
+
+    return this.users.length === 0 && this.editingUsers.length === 0;
+  },
+
+  _appendUser(users, user, attrs) {
+    let existingUser;
+    let usersLength = 0;
+
+    users.forEach(u => {
+      if (u.id === user.id) {
+        existingUser = u;
+      }
+
+      if (attrs && attrs.post_id) {
+        if (u.post_id === attrs.post_id) usersLength++;
+      } else {
+        usersLength++;
+      }
+    });
+
+    const props = attrs || {};
+    props.last_seen = Date.now();
+
+    if (existingUser) {
+      existingUser.setProperties(props);
+    } else {
+      const limit = this.get("siteSettings.presence_max_users_shown");
+
+      if (usersLength < limit) {
+        users.pushObject(EmberObject.create(Object.assign(user, props)));
```

Comment:
```suggestion
        users.pushObject(EmberObject.create({}, Object.assign(user, props)));
```

Otherwise you are actually mutating user here, not sure this is what you wanted to do.

Reply:
@jjaffeux The suggested change is still mutating the user with props. 😛

Reply:
Oh yes didn't put it at right position :+1:

Reply:
Mutating the `user` object is fine here because it is from the message's payload.